### ISSUE TITLE
When tuning hyperparameters, the best_model does not have the right parameters

### ISSUE THREAD

**Issue by tlienart** [alan-turing-institute/MLJ.jl#256]
While working on tutorials, it struck me that everytime I was fitting a tuned model, the best parameters would be the default parameters (this is also true from the examples in the docs as far as I could tell). Once or twice ok, every time, I doubt it. 

Here's an example:

```julia
using RDatasets, MLJ
iris = dataset("datasets", "iris")
y, X = unpack(iris, ==(:Species), colname -> true)
@load KNNClassifier
kc = KNNClassifier(K=100) # ludicrous default
r = range(kc, :K, lower=1, upper=100, scale=:linear)
tm = TunedModel(model=kc, resampling=CV(nfolds=3),
                tuning=Grid(resolution=5), ranges=r)
self_tuning_kc = machine(tm, X, y)
fit!(self_tuning_kc)
fitted_params(self_tuning_kc).best_model.K
```

The value returned is ...`100`. 

One possible reason is the measurements are all infinite (??)

```julia
evaluate(KNNClassifier(K=5), X, y)
```

returns

```
Evaluating over 6 folds: 100%[=========================] Time: 0:00:00
(measure = MLJBase.CrossEntropy[cross_entropy],
 measurement = [Inf],
 per_fold = Array{Float64,1}[[-0.0, -0.0, 0.24197568038124703, Inf, Inf, 0.306353196878611]],
 per_observation = Array{Array{Float64,1},1}[[[-0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0  …  -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0], [-0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0  …  -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0], [0.2231435513142097, -0.0, 0.2231435513142097, -0.0, 0.2231435513142097, -0.0, 0.5108256237659905, -0.0, -0.0, -0.0  …  -0.0, -0.0, -0.0, 0.916290731874155, -0.0, 1.6094379124341003, -0.0, 1.6094379124341003, 0.2231435513142097, -0.0], [-0.0, -0.0, 1.6094379124341003, -0.0, -0.0, -0.0, -0.0, -0.0, Inf, 0.2231435513142097  …  -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0, -0.0], [-0.0, 0.2231435513142097, -0.0, -0.0, -0.0, -0.0, Inf, -0.0, -0.0, -0.0  …  -0.0, -0.0, -0.0, -0.0, 0.916290731874155, -0.0, 0.2231435513142097, -0.0, 0.2231435513142097, -0.0], [-0.0, 1.6094379124341003, 1.6094379124341003, -0.0, -0.0, -0.0, -0.0, -0.0, 0.916290731874155, 0.2231435513142097  …  -0.0, 0.2231435513142097, 0.2231435513142097, -0.0, -0.0, -0.0, 0.5108256237659905, 0.2231435513142097, -0.0, 0.5108256237659905]]],)
```

**Comment by ablaom** [comment]
@tlienart Re your iris example mentioned in our recent call. I was not able to reproduce:

```julia
julia> using MLJ
julia> X, y = @load_iris
julia> model = @load DecisionTreeClassifier 
julia> r = range(model, :max_depth, lower=1, upper=5)
julia> tuned_model = TunedModel(model=model, ranges=[r, ], measure=cross_entropy)
julia> m = machine(tuned_model, X, y)
julia> fit!(m)
julia> fitted_params(m).best_model
DecisionTreeClassifier(pruning_purity = 1.0,
                       max_depth = 2,
                       min_samples_leaf = 1,
                       min_samples_split = 2,
                       min_purity_increase = 0.0,
                       n_subfeatures = 0,
                       display_depth = 5,
                       post_prune = false,
                       merge_purity_threshold = 0.9,
                       pdf_smoothing = 0.05,) @ 8…00

julia> report(m)
(parameter_names = ["max_depth"],
 parameter_scales = Symbol[:linear],
 parameter_values = Any[1; 2; … ; 4; 5],
 measurements = [2.41442, 0.635327, 1.12462, 0.942644, 0.942644],
 best_measurement = 0.6353268061118232,)
```

The second hyperparameter value is the optimal: max_depth = 2 .

**Comment by ablaom** [comment]
Further to the KNN example:

```julia
julia> report(self_tuning_kc)
(parameter_names = ["K"],
 parameter_scales = Symbol[:linear],
 parameter_values = Any[1; 26; … ; 75; 100],
 measurements = [Inf, Inf, Inf, Inf, Inf],
 best_measurement = Inf,)
```

A cross_entropy loss of `Inf` means the model was 100% certain of a particular class in the test when in fact the class was different. It is conceivable (although surprising for high K) that this is the case for each fold. Assuming KNN is working properly, and this is the really the case, then there is no bug with tuning. 



I suppose that, as a point of "good practice", if making probabilistic predictions one should add pdf smoothing to predictions if they are likely to be concentrated in one class. I added a smoothing hyperparameter to DecisionTreeClassifier for this reason (do @load DecistionTreeClassifier; ?DecisionTreeClassfier). With smoothing, infinite cross-entropy losses never occur. 

Here's what I get if I use a Brier score instead of cross-entropy:

```julia
function brier_score(d::UnivariateFinite, y)
    levels = classes(d)
    pvec = broadcast(pdf, d, levels)
    offset = 1 + sum(pvec.^2)
    return 2*pdf(d, y) - offset
end
brier_score(yhat::AbstractVector{<:UnivariateFinite}, y) =
    broadcast(brier_score, yhat, y)

import MLJBase
MLJBase.orientation(::typeof(brier_score)) = :score
MLJBase.reports_each_observation(::typeof(brier_score)) = true

using RDatasets, MLJ
iris = dataset("datasets", "iris")
y, X = unpack(iris, ==(:Species), colname -> true)
@load KNNClassifier
kc = KNNClassifier(K=100) # ludicrous default
r = range(kc, :K, lower=1, upper=100, scale=:linear)
tm = TunedModel(model=kc, resampling=CV(nfolds=6, shuffle=true),
                tuning=Grid(resolution=5), ranges=r, measure=brier_score)
self_tuning_kc = machine(tm, X, y)
fit!(self_tuning_kc)
fitted_params(self_tuning_kc).best_model.K
report(self_tuning_kc)
(parameter_names = ["K"],
 parameter_scales = Symbol[:linear],
 parameter_values = Any[1; 26; … ; 75; 100],
 measurements = [-0.106667, -0.213156, -0.595968, -0.841214, -0.871916],
 best_measurement = -0.10666666666666667,)

fitted_params(self_tuning_kc).best_model.K
1
```

**Comment by ablaom** [comment]
Another option is use a deterministic loss (misclassification_rate is built-in) and add `operation=predict_mode` to the `TunedModel` constructor (or `evaluate`, etc):

```julia
using RDatasets, MLJ
iris = dataset("datasets", "iris")
y, X = unpack(iris, ==(:Species), colname -> true)
@load KNNClassifier
kc = KNNClassifier(K=100) # ludicrous default
r = range(kc, :K, lower=1, upper=100, scale=:linear)
tm = TunedModel(model=kc, resampling=CV(nfolds=6, shuffle=true, rng=1),
                tuning=Grid(resolution=5), ranges=r,
                measure=misclassification_rate,
                operation=predict_mode)
self_tuning_kc = machine(tm, X, y)
fit!(self_tuning_kc)
fitted_params(self_tuning_kc).best_model.K
report(self_tuning_kc)

(parameter_names = ["K"],
 parameter_scales = Symbol[:linear],
 parameter_values = Any[1; 26; … ; 75; 100],
 measurements = [0.0533333, 0.146667, 0.506667, 1.0, 1.0],
 best_measurement = 0.05333333333333334,)

fitted_params(self_tuning_kc).best_model.K
1
```

**Comment by ablaom** [comment]
See also here: 

https://github.com/alan-turing-institute/MLJBase.jl/issues/53

**Comment by tlienart** [comment]
Thanks a lot for this  and glad to see that it was in fact working with the max depth!

**Comment by ablaom** [comment]
@tlienart Are you happy to close this?
